File: slam/lib/lasIO.mjs
// Минимальный самодостаточный LAS 1.4 читатель/писатель — НЕ общего
// назначения (в отличие от laspy в Python-версии): поддерживает ТОЛЬКО
// point data record format 6 и 7 (без доп. байт), потому что это
// единственные форматы, которые собственный пайплайн когда-либо
// производит/читает (decodeRaw пишет 6, colorize читает 6 → пишет 7,
// georeference/ppkCorrection читают и переписывают тот же формат, что
// пришёл на входе). Общий LAZ (сжатие) не поддерживается — сжатие всегда
// делается отдельным вызовом pdal.exe (тот же приём, что уже проверен для
// COPC-конвертации в этой сессии).
//
// Формат/офсеты — ASPRS LAS 1.4 spec (заголовок фиксированный 375 байт для
// версии 1.4, без VLR со своей CRS — координатная система, если нужна,
// прикладывается отдельным вызовом `pdal translate --add-dimension`/
// `pdal translate` с указанием `--writers.las.a_srs`, а не собственным VLR
// — проще и надёжнее, чем вручную собирать WKT VLR).

import { openSync, closeSync, readSync, writeSync, fstatSync, ftruncateSync } from "node:fs";

const HEADER_SIZE = 375;
const POINT_SIZE = { 6: 30, 7: 36 };

function writeHeader(fd, opts) {
  const { pointFormat, scale, offset, count, mins, maxs } = opts;
  const buf = Buffer.alloc(HEADER_SIZE);
  buf.write("LASF", 0, "ascii");
  buf.writeUInt16LE(0, 4); // file source ID
  buf.writeUInt16LE(1, 6); // global encoding: bit0 = standard GPS time
  // GUID (16 bytes) left zero
  buf.writeUInt8(1, 24); // version major
  buf.writeUInt8(4, 25); // version minor
  buf.write("OTHER", 26, "ascii");
  buf.write("gisdata-slam", 58, "ascii");
  buf.writeUInt16LE(0, 90); // creation day of year
  buf.writeUInt16LE(0, 92); // creation year
  buf.writeUInt16LE(HEADER_SIZE, 94);
  buf.writeUInt32LE(HEADER_SIZE, 96); // offset to point data (no VLRs)
  buf.writeUInt32LE(0, 100); // number of VLRs
  buf.writeUInt8(pointFormat, 104);
  buf.writeUInt16LE(POINT_SIZE[pointFormat], 105);
  buf.writeUInt32LE(count <= 0xffffffff ? count : 0, 107); // legacy count
  for (let i = 0; i < 5; i++) buf.writeUInt32LE(0, 111 + i * 4); // legacy points by return
  buf.writeDoubleLE(scale[0], 131);
  buf.writeDoubleLE(scale[1], 139);
  buf.writeDoubleLE(scale[2], 147);
  buf.writeDoubleLE(offset[0], 155);
  buf.writeDoubleLE(offset[1], 163);
  buf.writeDoubleLE(offset[2], 171);
  buf.writeDoubleLE(maxs[0], 179);
  buf.writeDoubleLE(mins[0], 187);
  buf.writeDoubleLE(maxs[1], 195);
  buf.writeDoubleLE(mins[1], 203);
  buf.writeDoubleLE(maxs[2], 211);
  buf.writeDoubleLE(mins[2], 219);
  // waveform/extended-VLR pointers (u64) — write as two u32 (Node has no
  // writeBigUInt64LE issues here, but plain 0 is simplest and correct).
  buf.writeBigUInt64LE(0n, 227);
  buf.writeBigUInt64LE(0n, 235);
  buf.writeUInt32LE(0, 243);
  buf.writeBigUInt64LE(BigInt(count), 247);
  for (let i = 0; i < 15; i++) buf.writeBigUInt64LE(0n, 255 + i * 8);
  writeSync(fd, buf, 0, HEADER_SIZE, 0);
}

/**
 * @param {string} path
 * @param {object} opts
 * @param {6|7} opts.pointFormat
 * @param {Float64Array} opts.x
 * @param {Float64Array} opts.y
 * @param {Float64Array} opts.z
 * @param {Float64Array} [opts.gpsTime]
 * @param {Uint16Array|Float64Array} [opts.intensity]
 * @param {Uint8Array} [opts.red] 8-bit per channel (scaled to 16-bit on write)
 * @param {Uint8Array} [opts.green]
 * @param {Uint8Array} [opts.blue]
 * @param {[number,number,number]} [opts.scale]
 */
export function writeLas(path, opts) {
  const { pointFormat, x, y, z } = opts;
  const n = x.length;
  if (![6, 7].includes(pointFormat)) throw new Error(`Unsupported point format ${pointFormat} (only 6/7)`);
  if (pointFormat === 7 && (!opts.red || !opts.green || !opts.blue)) {
    throw new Error("point format 7 requires red/green/blue");
  }

  let xMin = Infinity, yMin = Infinity, zMin = Infinity;
  let xMax = -Infinity, yMax = -Infinity, zMax = -Infinity;
  for (let i = 0; i < n; i++) {
    if (x[i] < xMin) xMin = x[i]; if (x[i] > xMax) xMax = x[i];
    if (y[i] < yMin) yMin = y[i]; if (y[i] > yMax) yMax = y[i];
    if (z[i] < zMin) zMin = z[i]; if (z[i] > zMax) zMax = z[i];
  }
  if (n === 0) { xMin = yMin = zMin = xMax = yMax = zMax = 0; }

  const scale = opts.scale || [0.001, 0.001, 0.001];
  const offset = [xMin, yMin, zMin];
  const pointSize = POINT_SIZE[pointFormat];

  const fd = openSync(path, "w");
  try {
    writeHeader(fd, { pointFormat, scale, offset, count: n, mins: [xMin, yMin, zMin], maxs: [xMax, yMax, zMax] });

    const CHUNK = 500_000;
    const buf = Buffer.alloc(pointSize * Math.min(CHUNK, Math.max(n, 1)));
    let filePos = HEADER_SIZE;
    for (let start = 0; start < n; start += CHUNK) {
      const end = Math.min(start + CHUNK, n);
      const cn = end - start;
      for (let i = 0; i < cn; i++) {
        const gi = start + i;
        const base = i * pointSize;
        buf.writeInt32LE(Math.round((x[gi] - offset[0]) / scale[0]), base);
        buf.writeInt32LE(Math.round((y[gi] - offset[1]) / scale[1]), base + 4);
        buf.writeInt32LE(Math.round((z[gi] - offset[2]) / scale[2]), base + 8);
        const intensity = opts.intensity ? Math.max(0, Math.min(65535, Math.round(opts.intensity[gi]))) : 0;
        buf.writeUInt16LE(intensity, base + 12);
        buf.writeUInt8(1, base + 14); // return number=1, number of returns=1 (packed: (1)|(1<<4))
        buf.writeUInt8(0, base + 15); // classification flags/channel/direction/edge
        buf.writeUInt8(0, base + 16); // classification
        buf.writeUInt8(0, base + 17); // user data
        buf.writeInt16LE(0, base + 18); // scan angle
        buf.writeUInt16LE(0, base + 20); // point source ID
        buf.writeDoubleLE(opts.gpsTime ? opts.gpsTime[gi] : 0, base + 22);
        if (pointFormat === 7) {
          buf.writeUInt16LE(opts.red[gi] * 257, base + 30);
          buf.writeUInt16LE(opts.green[gi] * 257, base + 32);
          buf.writeUInt16LE(opts.blue[gi] * 257, base + 34);
        }
      }
      writeSync(fd, buf, 0, cn * pointSize, filePos);
      filePos += cn * pointSize;
    }
    ftruncateSync(fd, filePos);
  } finally {
    closeSync(fd);
  }
}

/**
 * Reads an entire LAS (format 6/7) into memory as typed arrays. For files
 * produced by this pipeline's own steps only (see module docstring) —
 * arbitrary third-party LAS files with extra bytes/other formats are out of
 * scope.
 */
export function readLas(path) {
  const fd = openSync(path, "r");
  try {
    const header = Buffer.alloc(HEADER_SIZE);
    readSync(fd, header, 0, HEADER_SIZE, 0);
    if (header.toString("ascii", 0, 4) !== "LASF") throw new Error(`Not a LAS file: ${path}`);
    const pointFormat = header.readUInt8(104) & 0x7f; // top bit = "has WKT" flag in 1.4, ignore
    if (![6, 7].includes(pointFormat)) throw new Error(`Unsupported point format ${pointFormat} in ${path}`);
    const pointDataOffset = header.readUInt32LE(96);
    const pointSize = header.readUInt16LE(105);
    const scale = [header.readDoubleLE(131), header.readDoubleLE(139), header.readDoubleLE(147)];
    const offset = [header.readDoubleLE(155), header.readDoubleLE(163), header.readDoubleLE(171)];
    let count = Number(header.readBigUInt64LE(247));
    if (count === 0) count = header.readUInt32LE(107); // fall back to legacy count

    const x = new Float64Array(count);
    const y = new Float64Array(count);
    const z = new Float64Array(count);
    const intensity = new Float64Array(count);
    const gpsTime = new Float64Array(count);
    const red = pointFormat === 7 ? new Uint8Array(count) : null;
    const green = pointFormat === 7 ? new Uint8Array(count) : null;
    const blue = pointFormat === 7 ? new Uint8Array(count) : null;

    const CHUNK = 500_000;
    const buf = Buffer.alloc(pointSize * Math.min(CHUNK, Math.max(count, 1)));
    let filePos = pointDataOffset;
    for (let start = 0; start < count; start += CHUNK) {
      const end = Math.min(start + CHUNK, count);
      const cn = end - start;
      readSync(fd, buf, 0, cn * pointSize, filePos);
      filePos += cn * pointSize;
      for (let i = 0; i < cn; i++) {
        const gi = start + i;
        const base = i * pointSize;
        x[gi] = buf.readInt32LE(base) * scale[0] + offset[0];
        y[gi] = buf.readInt32LE(base + 4) * scale[1] + offset[1];
        z[gi] = buf.readInt32LE(base + 8) * scale[2] + offset[2];
        intensity[gi] = buf.readUInt16LE(base + 12);
        gpsTime[gi] = buf.readDoubleLE(base + 22);
        if (pointFormat === 7) {
          red[gi] = Math.round(buf.readUInt16LE(base + 30) / 257);
          green[gi] = Math.round(buf.readUInt16LE(base + 32) / 257);
          blue[gi] = Math.round(buf.readUInt16LE(base + 34) / 257);
        }
      }
    }
    return { pointFormat, count, scale, offset, x, y, z, intensity, gpsTime, red, green, blue };
  } finally {
    closeSync(fd);
  }
}

File: pyserver/slam_pipeline/pipeline/las_io.py
"""
Port of slam/lib/lasIO.mjs. Minimal, self-contained LAS 1.4 reader/writer --
NOT general-purpose (unlike Python's laspy): supports ONLY point data record
format 6 and 7 (no extra bytes), because that's the only pair this
pipeline's own steps ever produce/consume (decode_raw writes 6, colorize
reads 6 -> writes 7, georeference/ppk_correction read and rewrite whatever
format arrived on input). General LAZ (compression) is not supported --
compression is always a separate pdal call (see build_octree.py), same
pattern already used for COPC conversion elsewhere in this project.

Header/offsets follow the ASPRS LAS 1.4 spec (fixed 375-byte header for
version 1.4, no CRS VLR -- a coordinate system, if needed, is applied by a
separate `pdal translate --writers.las.a_srs`/`--writers.copc.a_srs` call
rather than hand-assembling a WKT VLR here -- simpler and more reliable
than building one by hand).

Vectorized with numpy structured arrays (the original .mjs looped byte-by-
byte since JS has no equivalent) -- same exact binary layout, much faster
for the multi-million-point clouds this pipeline handles.
"""

import struct

import numpy as np

HEADER_SIZE = 375
POINT_SIZE = {6: 30, 7: 36}

_DTYPE_6 = np.dtype([
    ("x", "<i4"), ("y", "<i4"), ("z", "<i4"),
    ("intensity", "<u2"),
    ("return_byte", "u1"), ("flags_byte", "u1"), ("classification", "u1"), ("user_data", "u1"),
    ("scan_angle", "<i2"), ("point_source_id", "<u2"),
    ("gps_time", "<f8"),
])
_DTYPE_7 = np.dtype(_DTYPE_6.descr + [("red", "<u2"), ("green", "<u2"), ("blue", "<u2")])
assert _DTYPE_6.itemsize == POINT_SIZE[6]
assert _DTYPE_7.itemsize == POINT_SIZE[7]


def _dtype_for(point_format: int) -> np.dtype:
    return _DTYPE_7 if point_format == 7 else _DTYPE_6


def _write_header(f, point_format, scale, offset, count, mins, maxs):
    buf = bytearray(HEADER_SIZE)
    buf[0:4] = b"LASF"
    struct.pack_into("<H", buf, 4, 0)  # file source ID
    struct.pack_into("<H", buf, 6, 1)  # global encoding: bit0 = standard GPS time
    # GUID (16 bytes) left zero
    buf[24] = 1  # version major
    buf[25] = 4  # version minor
    buf[26:26 + len(b"OTHER")] = b"OTHER"
    buf[58:58 + len(b"gisdata-slam")] = b"gisdata-slam"
    struct.pack_into("<H", buf, 90, 0)  # creation day of year
    struct.pack_into("<H", buf, 92, 0)  # creation year
    struct.pack_into("<H", buf, 94, HEADER_SIZE)
    struct.pack_into("<I", buf, 96, HEADER_SIZE)  # offset to point data (no VLRs)
    struct.pack_into("<I", buf, 100, 0)  # number of VLRs
    buf[104] = point_format
    struct.pack_into("<H", buf, 105, POINT_SIZE[point_format])
    struct.pack_into("<I", buf, 107, count if count <= 0xFFFFFFFF else 0)  # legacy count
    struct.pack_into("<5I", buf, 111, 0, 0, 0, 0, 0)  # legacy points by return
    struct.pack_into("<3d", buf, 131, *scale)
    struct.pack_into("<3d", buf, 155, *offset)
    struct.pack_into("<dddddd", buf, 179, maxs[0], mins[0], maxs[1], mins[1], maxs[2], mins[2])
    struct.pack_into("<Q", buf, 227, 0)  # start of waveform data packet record
    struct.pack_into("<Q", buf, 235, 0)  # start of first extended VLR
    struct.pack_into("<I", buf, 243, 0)  # number of extended VLRs
    struct.pack_into("<Q", buf, 247, count)
    struct.pack_into("<15Q", buf, 255, *([0] * 15))
    f.write(bytes(buf))


def write_las(path, point_format, x, y, z, gps_time=None, intensity=None, red=None, green=None, blue=None, scale=(0.001, 0.001, 0.001)):
    """
    x/y/z: array-like of float64 (metres). gps_time/intensity: array-like of
    float64, optional. red/green/blue: array-like of uint8 (0-255), required
    for point_format 7.
    """
    if point_format not in (6, 7):
        raise ValueError(f"Unsupported point format {point_format} (only 6/7)")
    if point_format == 7 and (red is None or green is None or blue is None):
        raise ValueError("point format 7 requires red/green/blue")

    x = np.asarray(x, dtype=np.float64)
    y = np.asarray(y, dtype=np.float64)
    z = np.asarray(z, dtype=np.float64)
    n = x.shape[0]

    if n:
        x_min, y_min, z_min = float(x.min()), float(y.min()), float(z.min())
        x_max, y_max, z_max = float(x.max()), float(y.max()), float(z.max())
    else:
        x_min = y_min = z_min = x_max = y_max = z_max = 0.0

    offset = (x_min, y_min, z_min)
    dtype = _dtype_for(point_format)
    points = np.zeros(n, dtype=dtype)
    points["x"] = np.round((x - offset[0]) / scale[0]).astype(np.int32)
    points["y"] = np.round((y - offset[1]) / scale[1]).astype(np.int32)
    points["z"] = np.round((z - offset[2]) / scale[2]).astype(np.int32)
    if intensity is not None:
        points["intensity"] = np.clip(np.round(np.asarray(intensity, dtype=np.float64)), 0, 65535).astype(np.uint16)
    points["return_byte"] = 1  # return number=1, number of returns=1 (packed, matches original byte-for-byte)
    if gps_time is not None:
        points["gps_time"] = np.asarray(gps_time, dtype=np.float64)
    if point_format == 7:
        points["red"] = np.asarray(red, dtype=np.uint16) * 257
        points["green"] = np.asarray(green, dtype=np.uint16) * 257
        points["blue"] = np.asarray(blue, dtype=np.uint16) * 257

    with open(path, "wb") as f:
        _write_header(f, point_format, scale, offset, n, (x_min, y_min, z_min), (x_max, y_max, z_max))
        f.write(points.tobytes())


class LasData:
    __slots__ = (
        "point_format", "count", "scale", "offset",
        "x", "y", "z", "intensity", "gps_time", "red", "green", "blue",
    )

    def __init__(self, **kwargs):
        for k, v in kwargs.items():
            setattr(self, k, v)


def read_las(path) -> LasData:
    """Reads an entire LAS (format 6/7) into memory as numpy arrays. For
    files produced by this pipeline's own steps only (see module docstring)
    -- arbitrary third-party LAS files with extra bytes/other formats are
    out of scope."""
    with open(path, "rb") as f:
        header = f.read(HEADER_SIZE)
        if header[0:4] != b"LASF":
            raise ValueError(f"Not a LAS file: {path}")
        point_format = header[104] & 0x7F  # top bit = "has WKT" flag in 1.4, ignore
        if point_format not in (6, 7):
            raise ValueError(f"Unsupported point format {point_format} in {path}")
        point_data_offset = struct.unpack_from("<I", header, 96)[0]
        point_size = struct.unpack_from("<H", header, 105)[0]
        scale = struct.unpack_from("<3d", header, 131)
        offset = struct.unpack_from("<3d", header, 155)
        (count,) = struct.unpack_from("<Q", header, 247)
        if count == 0:
            count = struct.unpack_from("<I", header, 107)[0]  # fall back to legacy count

        f.seek(point_data_offset)
        dtype = _dtype_for(point_format)
        raw = f.read(count * point_size)
        points = np.frombuffer(raw, dtype=dtype, count=count)

    x = points["x"].astype(np.float64) * scale[0] + offset[0]
    y = points["y"].astype(np.float64) * scale[1] + offset[1]
    z = points["z"].astype(np.float64) * scale[2] + offset[2]
    intensity = points["intensity"].astype(np.float64)
    gps_time = points["gps_time"].astype(np.float64)
    red = green = blue = None
    if point_format == 7:
        red = np.round(points["red"].astype(np.float64) / 257).astype(np.uint8)
        green = np.round(points["green"].astype(np.float64) / 257).astype(np.uint8)
        blue = np.round(points["blue"].astype(np.float64) / 257).astype(np.uint8)

    return LasData(
        point_format=point_format, count=count, scale=scale, offset=offset,
        x=x, y=y, z=z, intensity=intensity, gps_time=gps_time, red=red, green=green, blue=blue,
    )
